osu.Game/Scoring/ScoreInfo.cs

```diff
@@ -22,15 +22,29 @@
 
 namespace osu.Game.Scoring
 {
+    /// <summary>
+    /// A realm model containing metadata for a single score.
+    /// </summary>
     [ExcludeFromDynamicCompile]
     [MapTo("Score")]
     public class ScoreInfo : RealmObject, IHasGuidPrimaryKey, IHasRealmFiles, ISoftDelete, IEquatable<ScoreInfo>, IScoreInfo
     {
         [PrimaryKey]
         public Guid ID { get; set; }
 
+        /// <summary>
+        /// The <see cref="BeatmapInfo"/> this score was made against.
+        /// </summary>
+        /// <remarks>
+        /// When setting this, make sure to also set <see cref="BeatmapHash"/> to allow relational consistency when a beatmap is potentially changed.
+        /// </remarks>
```

Comment:
If anything scares me in this diff it's this. Having to manage both `BeatmapInfo` and `BeatmapHash` manually is mental overhead that we will have to carry forever from this point on. I'm looking at usages of `BeatmapInfo.set`, and some don't set the hash - for most of them it looks like it's because they don't really need to, like in this case:

https://github.com/ppy/osu/blob/4fdba880b1efc2dcb49a2f3c64585f2e896a64ba/osu.Game/Beatmaps/BeatmapImporter.cs#L86-L97

But there's also stuff like this:

https://github.com/ppy/osu/blob/4fdba880b1efc2dcb49a2f3c64585f2e896a64ba/osu.Game/Scoring/ScoreImporter.cs#L65-L69

which I don't think is a bug either because there the `BeatmapHash` gets populated when the score is parsed, but hopefully you get what I'm driving at that the mental overhead of keeping this second integrity check is nonzero? I don't trust that we will be catching this at review time.

This is a difficult problem and the approach taken here doesn't feel like the optimal one either necessarily. Dead records in storage are never pretty to deal with to the point where I would probably agree with solving this issue in a very explicit manner by prompting the user that updating/editing a beatmap will get their existing scores fully wiped (and then following up on that threat if they agree), just to keep our stuff clean. The behaviour proposed here is very "spooky action at a distance" with scores appearing and disappearing for reasons not entirely clear to the end user.

But if we are to go with this, then I guess at this point my main question would be - is it feasible (and would you agree) to have the setter of `BeatmapInfo` automatically modify `BeatmapHash` too (and maybe privatise the setter of `BeatmapHash`, if realm can deal with this)?

Reply:
>is it feasible (and would you agree) to have the setter of BeatmapInfo automatically modify BeatmapHash too

We can try to do smth like this (not tested tho):
```csharp
/// <summary>
/// The <see cref="BeatmapInfo"/> this score was made against.
/// </summary>
[Ignored]
public BeatmapInfo BeatmapInfo
{
    get => beatmapInfo;
    set
    {
        beatmapInfo = value;
        beatmapHash = value.Hash;
    }
}

/// <summary>
/// The <see cref="osu.Game.Beatmaps.BeatmapInfo.Hash"/> at the point in time when the score was set.
/// </summary>
[MapTo("BeatmapHash")]
private string beatmapHash { get; set; } = string.Empty;

[MapTo("BeatmapInfo")]
private BeatmapInfo beatmapInfo { get; set; } = null!;
```
Would be cool if it works

> Dead records in storage are never pretty to deal

I was relying on the osu!stable behavior. I don't have any source code, but just from user experience I can assume it have the same issue. The question is does dead scores really so much of a deal? Like my osu stable folder as a player always took a lot of space (compare to original size)
![image](https://user-images.githubusercontent.com/50776304/217676526-858cef7d-d8d4-4a71-9b4a-5ebb53bc6e55.png)

I just got used to it.

Alternative solutions I have in mind:
1. Add button `Clean-up dead scores` which is scanning for, removing dead scores
1. Wipe scores on every map update as you propose. With online score fetch it should be ok (tho we lose replays)

Reply:
Yeah before I make any final call in either direction I'd like to get more thoughts on what I said from @ppy/team-client.

Reply:
I'm okay with the setter implementation. Though I believe realm will only work with `public` properties, so that would need to be considered (you can see other examples like:

https://github.com/ppy/osu/blob/27c497145f87449ac17c08796fc57026d3129f2e/osu.Game/Scoring/ScoreInfo.cs#L233-L257

Reply:
> I believe realm will only work with public properties

Not sure about it. According to my testing it works with private **properties** (doesn't work with private *fields* tho).
![image](https://user-images.githubusercontent.com/50776304/218097976-2a13d57d-0bfb-4331-b653-7978f82b9440.png)